contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,985 @@
+     */
+    function issueExactSetForLongToken(
+        ISetToken _setToken,
+        uint256 _amountSetToken,
+        uint256 _maxAmountInputToken,
+        Exchange _exchange
+    )
+        isSetToken(_setToken)
+        external
+        nonReentrant
+    {
+        bytes memory paymentParams = abi.encode(_maxAmountInputToken);
+        initiateIssuance(_setToken, _amountSetToken, _exchange, PaymentToken.LongToken, paymentParams);
+    }
+
+    function issueExactSetForERC20(
+        ISetToken _setToken,
+        uint256 _amountSetToken,
+        address _inputToken,
+        uint256 _maxAmountInputToken,
+        Exchange _exchange
+    )
+        isSetToken(_setToken)
+        external
+        nonReentrant
+    {
+        bytes memory paymentParams = abi.encode(_maxAmountInputToken, _inputToken);
+        initiateIssuance(_setToken, _amountSetToken, _exchange, PaymentToken.ERC20, paymentParams);
+    }
+
+    /**
+     * Trigger issuance of set token paying with the underlying of the collateral token directly
+     *
+     * @param _setToken            Set token to issue
+     * @param _amountSetToken      Amount to issue
+     * @param _exchange            Exchange to use in swap from short to long token
+     */
+    function issueExactSetForETH(
+        ISetToken _setToken,
+        uint256 _amountSetToken,
+        Exchange _exchange
+    )
+        isSetToken(_setToken)
+        external
+        payable
+        nonReentrant
+    {
+        bytes memory paymentParams = abi.encode(msg.value);
+        initiateIssuance(_setToken, _amountSetToken, _exchange, PaymentToken.ETH, paymentParams);
+    }
+
+    /**
+     * This is the callback function that will be called by the AaveLending Pool after flashloaned tokens have been sent
+     * to this contract.
+     * After exiting this function the Lending Pool will attempt to transfer back the loaned tokens + interest. If it fails to do so
+     * the whole transaction gets reverted
+     *
+     * @param assets     Addresses of all assets that were borrowed
+     * @param amounts    Amounts that were borrowed
+     * @param premiums   Interest to be paid on top of borrowed amount
+     * @param params     Encoded bytestring of other parameters from the original contract call to be used downstream
+     */
+    function executeOperation(
+        address[] calldata assets,
+        uint256[] calldata amounts,
+        uint256[] calldata premiums,
+        address , // 
+        bytes calldata params
+    ) external override  onlyLendingPool returns (bool) {
+        require(assets.length == 1, "Exchange Issuance Leveraged: TOO MANY ASSETS");
+        require(amounts.length == 1, "Exchange Issuance Leveraged: TOO MANY AMOUNTS");
+        require(premiums.length == 1, "Exchange Issuance Leveraged: TOO MANY PREMIUMS");
+
+        (,,,bool isIssuance,,,) = _decodeParams(params);
+        if(isIssuance){
+            _depositLongToken(assets[0], amounts[0]);
+            _issueSet(params);
```

Comment:
A lot of these functions take `params` but don't use most of the parameters. Can we instead have them take the exact parameters they need? Being able to look at what parameters go into each functions helps a lot with readability.

Reply:
I did do that initially but I ended up with quite a few "stack to deep" errors when I decoded all of the params top level and handed each function the params it needed.

That's why I ended up just passing the params through.

Reply:
Do you see any other way to avoid the "stack too deep" ? Obviously the current approach beside reducing readability is also gas inefficient since we decode the params more than once, but I wasn't able to make it work yet with only decoding once due to above. (Too many variables)

Reply:
So there are two main ways you can avoid this. First, you can create a new scope by using a `{}` and putting chunks of code between it. This may not work though since you might just have too many variables floating around anyway even if you are clever about placing brackets to create new stack frames. Another solution to this is to have your `_decodeParams` helper return a struct. You can then still do `someFunc(ourStruct.param1, outStruct.param2)` without loading every other variable in the struct into the stack.

Reply:
Nice, the struct approach sounds pretty straight forward so I'll try that first. Thanks a lot.

Reply:
Implemented the struct approach. seems to work fine